old scheduler criticality recursion taken out

jvansomeren opened the pull request.

jvansomeren committed:
Scheduler::criticality_lessthan simplified by taking recursion out
LOG_IF_DEBUGs mostly disabled

Scheduler::criticality_lessthan simplified by taking recursion out:
when the criticality of two nodes is equal, the size of the lists of depending nodes decides: the longer, the more critical
otherwise, the original order is used (when earlier than more critical)
when global option scheduler_heuristic == "path_length" the above heuristic is taken
when global option scheduler_heuristic == "random" only the criticality value is taken, not the size of lists of dep nodes
Please note: ListScheduler was not updated with this (wait for Scheduler to become stable).

LOG_IF_DEBUGs mostly disabled:
That output poluted the text too much and overview was lost; please enable again when needed.

jvansomeren committed:
Only more QL_DOUT lines for debugging various gate creation and decomposition schemes

jvansomeren committed:
Disabled LOG_DEBUG output of IR etc. while converting between old and new

jvansomeren committed:
libqasm change (extern) and QL_DOUTs in future.cc and past.cc

jvansomeren committed:
consequence of updated scheduler strategy (criticality) to golden files

jvansomeren committed:
dummy commit

wvlothuizen commented on src/ql/ir/compat/kernel.cc:
```diff
@@ -553,8 +553,10 @@ Bool Kernel::add_custom_gate_if_available(
 
     // first check if a specialized custom gate is available
     // a specialized custom gate is of the form: "cz q0 q3"
+    QL_DOUT("is specialized custom gate availablefor for instr " << instr);
```
typo

wvlothuizen commented on src/ql/ir/compat/platform.cc:
```diff
@@ -460,6 +460,7 @@ void Platform::load(
     const utils::Str &platform_config_fname,
     const utils::Str &compiler_config
 ) {
+    QL_DOUT("compat::platform::load");
```
"compat::platform" seems a bit redundant since QL_DOUT already prints full file name

wvlothuizen commented on src/ql/ir/new_to_old.cc:
```diff
@@ -801,18 +801,30 @@ NewToOldConverter::NewToOldConverter(const Ref &ir) : ir(ir) {
     }
+    QL_DOUT("NewToOldConvertor old instruction_map:");
+    QL_DOUT("(disabled)");
+//    QL_IF_LOG_DEBUG {
```
use #if/#else instead of comments to choose between disabled and active. Does this have generic use?

wvlothuizen commented on src/ql/ir/old_to_new.cc:
```diff
@@ -921,7 +921,8 @@ Ref convert_old_to_new(const compat::PlatformRef &old) {
     // Check the result.
     QL_DOUT("Result of old->new IR platform conversion:");
-    QL_IF_LOG_DEBUG(ir->dump_seq());
+    QL_DOUT("(disabled)");
```
use #if/#else. I guess what we really need is finer control over log output per 'module' like Python has

jvansomeren replied:
Yes, agreed.
Don't know solution yet for this.
Updated the idiom to simulate enable/disable per file.
This allows finer control over endless dumping, at least
because the debug output got clogged up by those dumps.

wvlothuizen commented on src/ql/ir/old_to_new.cc:
```diff
@@ -1687,8 +1691,10 @@ Ref convert_old_to_new(const compat::ProgramRef &old) {
     QL_DOUT("Result of old->new IR program conversion:");
-    QL_IF_LOG_DEBUG(ir->dump_seq());
+    QL_DOUT("(disabled)");
+    // QL_IF_LOG_DEBUG(ir->dump_seq());
```
see above

wvlothuizen commented on src/ql/resource/instrument.cc:
```diff
@@ -464,33 +464,33 @@ void InstrumentResource::on_initialize(rmgr::Direction direction) {
-        std::cout << "nq_instr_qn: " << config->multi_qubit_instrument[2] << std::endl;
-        std::cout << "====================================" << std::endl;
-    }
+//    QL_IF_LOG_DEBUG {
```
disable using #if, not comments

wvlothuizen commented on src/ql/resource/inter_core_channel.cc:
```diff
@@ -185,26 +185,26 @@ void InterCoreChannelResource::on_initialize(rmgr::Direction direction) {
-        std::cout << "num_channels: " << config->num_channels << std::endl;
-        std::cout << "====================================" << std::endl;
-    }
+//    QL_IF_LOG_DEBUG {
```
see above

jvansomeren replied:
Updated the pattern of QL_IF_LOG_DEBUG use and definition.

jvansomeren committed:
QL_IF_LOG_DEBUG use adapted to allow file wise disabling of long ir dumps

Idiomatic use of QL_IF_LOG_DEBUG:

QL_IF_LOG_DEBUG {
    QL_DOUT("text that it will dump");
    dump code possibly not predicated by LOG_DEBUG
} else {
    QL_DOUT("text that it will dump (disabled)");
}

At the top of a file:

dumping enabled:
// comment lines below out to enable LOG_DEBUG ir dumping
/*#undef QL_IF_LOG_DEBUG
When log_level==LOG_DEBUG then dump is done
When log_level!=LOG_DEBUG nothing is printed

dumping disabled:
// comment lines below out to enable LOG_DEBUG ir dumping
When log_level==LOG_DEBUG then dump is not done but text ending in (disabled) is printed
When log_level!=LOG_DEBUG nothing is printed

jvansomeren committed:
Some QL_DOUTs updated

wvlothuizen commented on src/ql/com/topology.cc:
```diff
@@ -7,6 +7,10 @@
 
 #include "ql/utils/logger.h"
 
+// comment two lines below out to enable LOG_DEBUG ir dumping
```
Redefining a 'system' macro isn't good practice in my opinion. For one thing, the name no longer reflects the functionality, and although the compiler will probably remove unreachable code, it provides a runtime solution for a compile time choice.

I think it's cleaner to use:
```
#define DEBUG_VERBOSE 1

#if DEBUG_VERBOSE
...
#else
...
#endif
```

jvansomeren replied:
Yes, you're right.
Still the current solution (with #ifdef, see the commit message) doesn't seem right either.
I would prefer (a family of) options to turn this on/off per file (not per class or whatever because that is not obvious to the debugger), and to revise the QL_*OUT with that.
In CoSy we had a standard 'engine' (is pass here) option called 'debug' which has a level as value (similar to log_level) which requires QL_DOUT to be sensitive to a pass-level variable set by this option. The latter requires C++ expertise to implement here. The pass level is not appropriate in OpenQL because there is so much common code (not in pass) although we could interpret 'pass' dynamically, i.e. QL_DOUT encountered between the start and stop of a particular pass, which would include the common code too when it is executed on behalf of that pass.
This should be a separate design and not pursued in this current PR.

wvlothuizen commented on src/ql/pass/sch/list_schedule/list_schedule.cc:
```diff
@@ -210,9 +214,11 @@ void ListSchedulePass::run_on_block(
     // Always dump dot for the schedule if we're debugging.
-    if (QL_IS_LOG_DEBUG) {
-        QL_COUT("dumping dot file...");
+    QL_IF_LOG_DEBUG {
```
This makes the decision compile time instead of runtime

wvlothuizen commented on src/ql/resource/instrument.cc:
```diff
@@ -5,10 +5,11 @@
 #include "ql/resource/instrument.h"
 
+// comment lines below out to enable LOG_DEBUG ir dumping
 /*#undef QL_DOUT
```
remove orignal hack

jvansomeren committed:
Replaced QL_IF_LOG_DEBUG idiom use by MULTI_LINE_LOG_DEBUG idiom to control multi-line debugging output

New idiom:

// #define MULTI_LINE_LOG_DEBUG to enable multi-line dumping

...

    QL_IF_LOG_DEBUG {
        QL_DOUT("Result of old->new IR platform conversion:");
        ir->dump_seq();
    }
    QL_DOUT("Result of old->new IR platform conversion (disabled)");

In the above, 'ir->dump_seq()' is not guarded internally by LOG_DEBUG/QL_DOUT
and prints the full IR to the debugging log.
To signal that this printing is prepared for but is currently disabled,
the message after the #else is given.

What we really want is a family of options to turn this on/off per file.
Per module (??) or per class is too cumbersome.

jvansomeren commented:
Yes, you're right with your comments, hence a new idiom ...

Still the current solution (with #ifdef, see the commit message) doesn't seem right either.
I would prefer (a family of) options to turn this on/off per file (not per class or whatever because that is not obvious to the debugger), and to revise the QL_*OUT with that.
In CoSy we had a standard 'engine' (is pass here) option called 'debug' which has a level as value (similar to log_level) which requires QL_DOUT to be sensitive to a pass-level variable set by this option. The latter requires C++ expertise to implement here. The pass level is not appropriate in OpenQL because there is so much common code (not in pass) although we could interpret 'pass' dynamically, i.e. QL_DOUT encountered between the start and stop of a particular pass, which would include the common code too when it is executed on behalf of that pass.
This should be a separate design and not pursued in this current PR.

jvansomeren committed:
Merge branch 'develop' into mapper

wvlothuizen commented on src/ql/com/topology.cc:
```diff
@@ -7,6 +7,9 @@
 #include "ql/utils/logger.h"
 
+// #define MULTI_LINE_LOG_DEBUG to enable multi-line dumping 
+#undef MULTI_LINE_LOG_DEBUG
```
No need to undef, since there's an assumption this has file scope. I'll file an issue t extend the logging system with finer control at run time

jvansomeren replied:
Ok, I'll update this, commit and merge.
